pkg/handlers/handlers.go

```diff
@@ -30,58 +31,101 @@ import (
+	return mimeTypeReader{
+		mimeExt:  r.mime.Extension(),
+		mimeName: mimeType(r.mime.String()),
+		Reader:   r.BufferedReaderSeeker,
+	}
+}
+
+// newMimeTypeReader creates a new mimeTypeReader from an io.Reader.
+// It uses a bufio.Reader to perform MIME type detection on the input reader
+// without consuming it, by peeking into the first 512 bytes of the input.
+// This encapsulates both the original reader and the detected MIME type information.
+// This function is particularly useful for specialized archive handlers
+// that need to pass extracted content to the default handler without modifying the original reader.
+func newMimeTypeReader(r io.Reader) (mimeTypeReader, error) {
+	const defaultMinBufferSize = 3072
+	bufReader := bufio.NewReaderSize(r, defaultMinBufferSize)
+	// A buffer of 512 bytes is used since many file formats store their magic numbers within the first 512 bytes.
+	// If fewer bytes are read, MIME type detection may still succeed.
+	buffer, err := bufReader.Peek(defaultMinBufferSize)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return mimeTypeReader{}, fmt.Errorf("unable to read file for MIME type detection: %w", err)
+	}
+
+	mime := mimetype.Detect(buffer)
+
+	return mimeTypeReader{mimeExt: mime.Extension(), mimeName: mimeType(mime.String()), Reader: bufReader}, nil
+}
+
 // newFileReader creates a fileReader from an io.Reader, optionally using BufferedFileWriter for certain formats.
 func newFileReader(r io.Reader) (fileReader, error) {
-	var reader fileReader
+	var fReader fileReader
 
-	bufReader := iobuf.NewBufferedReaderSeeker(r)
+	fReader.BufferedReaderSeeker = iobuf.NewBufferedReaderSeeker(r)
 
-	mime, err := mimetype.DetectReader(bufReader)
+	// Disable buffering after initial reads.
+	// This optimization ensures we don't continue writing to the buffer after the initial reads.
+	defer fReader.DisableBuffering()
+
+	mime, err := mimetype.DetectReader(fReader)
 	if err != nil {
-		return reader, fmt.Errorf("unable to detect MIME type: %w", err)
+		return fReader, fmt.Errorf("unable to detect MIME type: %w", err)
 	}
-	reader.mimeType = mimeType(mime.String())
+	fReader.mime = mime
 
 	// Reset the reader to the beginning because DetectReader consumes the reader.
-	if _, err := bufReader.Seek(0, io.SeekStart); err != nil {
-		return reader, fmt.Errorf("error resetting reader after MIME detection: %w", err)
+	if _, err := fReader.Seek(0, io.SeekStart); err != nil {
+		return fReader, fmt.Errorf("error resetting reader after MIME detection: %w", err)
 	}
 
-	format, _, err := archiver.Identify("", bufReader)
+	// Check if the MIME type should bypass the archiver library identification and continue processing.
+	// This bypass is necessary for archive formats not supported by the archiver library and for text-based formats
+	// that are best handled directly by the default reader.
+	if _, ok := skipArchiverMimeTypes[mimeType(mime.String())]; ok {
+		return fReader, nil
+	}
```

Comment:
This was a further optimization for files we know don't need the archiver's `Identify` function. `Identify` reads (potentially from disk) a certain amount of data for identification, which can slow down processing for files on slower storage mediums. Detecting the MIME type before identification saves this additional read penalty. We only incur this penalty for archive types that must be handled by the archiver library, as we depend on the `format` returned by Identify in the archive handler.

Reply:
It does, I was just trying to avoid reading the data twice: once with `mimetype.DetectReader` and then again with `Identify`. By using a map lookup, we can avoid the second read for certain mime types.

Apologies for the poorly worded comment. The optimization aims to avoid the overhead of an additional read, not to ensure correctness. We could skip the map lookup and read the data twice; it would still function correctly.

Reply:
Ok, I think I'm caught up. I really recommend removing the word "necessary" from the comment, though - I think other, future readers could easily repeat my mistake here. Maybe something like: `If a MIME type is known to not be an archive type, we might as well return here rather than paying the I/O penalty of an archiver.Identify() call that won't identify anything.`